### PR TITLE
Koban token

### PR DESCRIPTION
**Pull request by VSBDev**
# Description

Kinto Koban Token as a tokenization template

## Type of change

- [ ] Bug fix 
- [X] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Deployment
- [ ] Forge Script
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [X] The diff is legible and has no extraneous changes
- [X] Complex code has been commented, including external interfaces
- [X] Tests have 100% code coverage
- [X] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

File: src/tokens/KintoKoban.sol [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        address toOwner = IKintoWallet(to).owners(0); // Owner of the KintoID for recipient
+        // Use kintoID for KYC checks
+        if (!kintoID.isKYC(fromOwner) || !kintoID.isKYC(toOwner)) {
+            return 1; // Either sender or recipient is not KYCed
+        }
+        // Check if the sender and recipient have the required country traits
+        bool fromFlagged = false;
+        bool toFlagged = false;
+        for (uint256 i = 0; i < countryList.length; i++) {
+            if (kintoID.hasTrait(fromOwner, uint16(countryList[i]))) {
```
Comment (ylv-io) [inline src/tokens/KintoKoban.sol:119]:
```suggestion
            if (!fromFlagged && kintoID.hasTrait(fromOwner, uint16(countryList[i]))) {
```

File: src/tokens/KintoKoban.sol [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                fromFlagged = true; // Sender has an allowed trait
+            }
+            if (kintoID.hasTrait(toOwner, uint16(countryList[i]))) {
```
Comment (ylv-io) [inline src/tokens/KintoKoban.sol:122]:
```suggestion
            if (!toFlagged && kintoID.hasTrait(toOwner, uint16(countryList[i]))) {
```

File: src/tokens/KintoKoban.sol [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+     * @notice Detects transfer restrictions based on KYC status, country list, and transfer amount
+     * @param from Address of the sender
+     * @param to Address of the recipient
+     * @param value Amount of tokens to transfer
+     * @return uint8 restriction code (0=success, 1=not KYCed, 2=country restriction, 3=exceeds transfer limit)
+     */
+    function detectTransferRestriction(address from, address to, uint256 value) public view returns (uint8) {
+        // Check for total transfer limit
+        if (value > TOTAL_TRANSFER_LIMIT) {
+            return 3; // Exceeds transfer limit
```
Comment (ylv-io) [inline src/tokens/KintoKoban.sol:105]: Let's define constants for these magic numbers
```
uint256 constant EXCEED_TRANFER_LIMIT = 3;
```

File: src/tokens/KintoKoban.sol [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        //use walletFactory to check they are a wallet walletTs > 0
+        if (WALLET_FACTORY.walletTs(from) == 0 || WALLET_FACTORY.walletTs(to) == 0) {
+            return NOT_KYCED_CODE; // Either sender or recipient is not a wallet
+        }
+
+        // Get the owner of the KintoID NFT for both wallets
+        address fromOwner = IKintoWallet(from).owners(0); // Owner of the KintoID for sender
+        address toOwner = IKintoWallet(to).owners(0); // Owner of the KintoID for recipient
+        // Use kintoID for KYC checks
+        if (!KINTO_ID.isKYC(fromOwner) || !KINTO_ID.isKYC(toOwner)) {
```
Comment (rrecuero) [inline src/tokens/KintoKoban.sol:121]: this is not needed. Everyone is kyced

**Comment by rrecuero** [comment]
My main comment is that this is more an example that a standard. Take a look at the base ERC20 contracts or the Vault EIP. This should be a baseclass that has the functions and the user can use it by inheriting and extending

**Pull request closed without merging**